Polio 683 : Vaccine tab improvement

quang-le opened the pull request:
In the vaccine managment tab, saving with an empty Shipment or Destruction would return a 400 and crash the page. Fixing this required fixing some other issues, so there's been some scope creep here

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [ ] Did I add translations
- [ ] My migrations file are included
- [ ] Are there enough tests

## Changes

- Change the validation in `useFormValidator` so that empty `Destruction` and  `Shipment` will block form save
- Improved validation for `Destruction`, `FormA` and `Shipment`. All empty fields will error if one field has a value, thus showing that all fields have to be filled out together (bar `comments`).  `FormA` is a special case since the form itself can't be deleted, as shipments and destrcutions can, so there's some custom code to handle that as well (mainly making sure the payload sent to the back end is correct).
- Add `SingleSelect` input component for better error management (will not be errored by default)
- Add `DebouncedTextInput` to replace `TextInput` in the Vaccine tab to avoid input lag
- Refactor `VaccineManagementForm` to avoid looping through all rounds and improve performance
- Add `disabled `prop`to `TextArea`
- Prevent commenting Form A , destruction, shipment if all mandatory fields are not filled by disabling the input



## How to test

Go to polio campaigns. try editing the vaccine management in a campaign, then try creating a new campaign with vaccine management data.

## Print screen / video

https://user-images.githubusercontent.com/38907762/206246485-75d062dd-a807-4cc2-b2b7-107b2903e2b3.mov

## Note

- I used `DebouncedInput` i.o. updating the existing (custom) `TextInput` of polio because adding the debounce is messing with the validation: since mandatory empty fields are in error, but we only show the error when the field has been `touched`, adding a debounce was messing with  that because of its reliance on `useEffect`. 
Since we plan to refactor the whole form and its API anyway, it seemed more work than was worth.

- TextArea should be updated to:
    - Disable onFocus styling when disabled
    - Be able to receive a helper text (use case: explain the user why the field is disabled)
   I didn't add it to limit scope creep

beygorghor commented:
<img width="1582" alt="Screenshot 2021-06-28 at 09 14 24" src="https://user-images.githubusercontent.com/12494624/123595608-65351580-d7f1-11eb-9f14-7874b6243615.png">
if you empty the field, the ui is displaying "NaN" and console is not happy.

olethanh commented:
Added a fix for this. There was also a different comportement for longitude than latitude so I also fixed that.

quang-le committed:
POLIO-683:prevent saving empty shipment,destruction

quang-le committed:
POLIO-683: add touchOnFocus prop to TextInput

quang-le committed:
POLIO-683: add SingleSelect for forms

quang-le committed:
POLIO-683: replaceSelect with SingleSelect

quang-le committed:
POLIO-683: improve error mgt for vaccine tab

- Shipments, Destructions: filling one field will error the others

quang-le committed:
POLIO-683: improve formA error state

- if one field has value, other error
- if all fields empty, error state removed

quang-le committed:
POLIO-683: fix  payload issue with formA

- send null i.o '' when deleting value

quang-le committed:
POLIO-683: improve vaccine tab render time

quang-le committed:
POLIO-683: WIp add debounce to polio TextInput

quang-le committed:
POLIO-683: remove debounce from TextInput

quang-le committed:
POLIO-683: fix type error in DebouncedTextInput

quang-le committed:
POLIO-683: remove looping in VaccineManagementForm

- improve performance by on ly rendering the displayed round

quang-le committed:
POLIO-683: use DebouncedTextINput in Vaccine tab

quang-le committed:
POLIO-683: fix bug with DebouncedText

- set initial value of control Ref to avoid overwrite with empty string

quang-le committed:
POLIO-683: improve error message of validator

quang-le committed:
POLIO-683: add disabled to TaxtArea

quang-le committed:
POLIO-683 disable comment if other fields empty

- For FormA, Destruction and Shipment

beygorghor commented:
Some aesthetical remarks:
- spacing between rows is too big :
![Screenshot 2022-12-09 at 16 12 07](https://user-images.githubusercontent.com/12494624/206733501-d0d7a41e-a9da-4756-acfd-255339e66082.png)
- comment textArea is using a colored background (disabled mode) and label has a white background:
![Screenshot 2022-12-09 at 16 12 22](https://user-images.githubusercontent.com/12494624/206733642-5a3aacb2-26b3-4554-b3ad-2c01474cad2e.png)
- not sure a `Divider` is required here:
![Screenshot 2022-12-09 at 16 15 46](https://user-images.githubusercontent.com/12494624/206733902-92c86de4-a997-4f8e-bd04-3fdd2da53a2d.png)
- i'm not convinced by putting all shipments or destruction fields on error after you update the first one
![Screenshot 2022-12-09 at 16 20 33](https://user-images.githubusercontent.com/12494624/206734656-3e00aad1-a3d2-4d53-a469-fdb6185f1f25.png)

beygorghor commented on hat/assets/js/apps/Iaso/components/forms/TextArea.tsx:
```diff
@@ -122,6 +124,7 @@ export const TextArea: FunctionComponent<Props> = ({
                     setTextValue(e.target.value);
                 }}
                 value={textValue}
+                disabled={disabled}
```
You need to apply specific style on the label if disabled and remove the withe background

quang-le replied:
I made a separate ticket for this

quang-le replied:
I actually fixed it and will close the additional ticket

beygorghor commented on plugins/polio/js/src/components/Inputs/DebouncedTextInput.tsx:
```diff
@@ -0,0 +1,103 @@
+    debounceTime = 0,
+    withMarginTop = false,
+    clearable = false,
+    // touchOnFocus = true,
```
you can remove this

beygorghor commented on plugins/polio/js/src/components/Inputs/DebouncedTextInput.tsx:
```diff
@@ -0,0 +1,103 @@
+    const [textValue, setTextValue] = useState(parsedValue);
+    const [debouncedValue] = useDebounce(textValue, debounceTime);
+
+    useEffect(() => {
+        prevDebounced.current = parsedValue;
+        // leaving the deps empty because the effect should only run once
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
```
`useEffect(() => {
        prevDebounced.current = parsedValue;
    });` will do the trick then !

quang-le replied:
Nope, because then the effect will be run at every render

beygorghor commented on plugins/polio/js/src/components/Inputs/TextInput.js:
```diff
@@ -17,7 +32,13 @@ export const TextInput = ({ field = {}, form = {}, value, ...props } = {}) => {
+            onFocus={
+                touchOnFocus
+                    ? () => form.setFieldTouched(field.name, true)
+                    : undefined
```
On Focus is waiting for a function, maybe return` () => null` ?

beygorghor commented on plugins/polio/js/src/forms/DestructionForm.tsx:
```diff
@@ -1,29 +1,64 @@
+    useEffect(() => {
+        // Using every to be able to break the loop
+        destructionFieldNames.every(key => {
+            if (fieldValues[key]) {
```
`fieldValues` can be `undefined` no ?

beygorghor commented on plugins/polio/js/src/forms/FormAForm.tsx:
```diff
@@ -0,0 +1,159 @@
+        [roundIndex, values?.rounds],
+    );
+
+    // // Make all fields error by setting touched to true if any other formA field has a value
```
i'm not sure of this behaviour

quang-le replied:
The requirement is: the fields can be empty, but if one has a value, then all should have a value (except comments)

beygorghor commented on plugins/polio/js/src/forms/FormAForm.tsx:
```diff
@@ -0,0 +1,159 @@
+        formAFieldNames.every(key => {
+            if (fieldValues?.[key]) {
+                formAFieldNames.forEach(name => {
+                    setFieldTouched(`${accessor}.${name}`, true);
+                });
+                // break the loop if any field has a value
+                return false;
+            }
+            return true;
+        });
+    }, [accessor, fieldValues, setFieldTouched]);
+
+    // // Remove error state if no field has value
+    useEffect(() => {
+        const isFormAEmpty = formAFieldNames.every(key => !fieldValues?.[key]);
+        if (isFormAEmpty) {
+            formAFieldNames.forEach(key => {
+                setFieldTouched(`${accessor}.${key}`, false);
+            });
+        }
+    }, [accessor, fieldValues, setFieldTouched]);
+
+    // // Set TextFields values to null if empty to avoid 400.
+    useEffect(() => {
+        if (fieldValues?.forma_usable_vials === '') {
+            setFieldValue(`${accessor}.forma_usable_vials`, null);
+        }
+        if (fieldValues?.forma_unusable_vials === '') {
+            setFieldValue(`${accessor}.forma_unusable_vials`, null);
+        }
+        if (fieldValues?.forma_missing_vials === '') {
+            setFieldValue(`${accessor}.forma_missing_vials`, null);
+        }
+    }, [
+        accessor,
+        fieldValues?.forma_missing_vials,
+        fieldValues?.forma_unusable_vials,
+        fieldValues?.forma_usable_vials,
+        setFieldValue,
+    ]);
```
Lot's of `usEffects` !  😨

quang-le replied:
I know. The last one is the worst, but without it we get a 400 from the backend

beygorghor commented on plugins/polio/js/src/forms/FormAForm.tsx:
```diff
@@ -0,0 +1,159 @@
+                spacing={2}
+                justifyContent="flex-start"
+            >
+                <Grid item lg={3} md={6}>
```
this is why you have too much spacing, try to put `forma_date` under `forma_reception` , `forma_comment ` under `formAUnusableVials `

beygorghor commented on plugins/polio/js/src/forms/ShipmentForm.tsx:
```diff
@@ -1,23 +1,71 @@
+    useEffect(() => {
+        // Using every to be able to break the loop
+        shipmentFieldNames.every(key => {
+            if (fieldValues[key]) {
```
same as before, `fieldValues ` can be `undefined`

beygorghor replied:
Also you have the same spacing issue here

quang-le replied:
It comes from the default padding from the `spacing` prop. We need to override it, either by dividing both padding top and by removing one entirely.

quang-le replied:
removing the padding doesn't solve the problem. I think the inputs's height might interfere here. Leaving as is for now

beygorghor commented on plugins/polio/js/src/forms/VaccineManagementForm.tsx:
```diff
@@ -57,189 +55,119 @@ export const VaccineManagementForm: FunctionComponent<Props> = () => {
                         <Box mb={2}>
                             <Tabs
-                                value={currentRoundNumber}
+                                // value={currentRoundNumber}
```
to be removed

beygorghor commented on plugins/polio/js/src/forms/VaccineManagementForm.tsx:
```diff
@@ -57,189 +55,119 @@ export const VaccineManagementForm: FunctionComponent<Props> = () => {
                                     <Tab
-                                        key={round.number}
+                                        // eslint-disable-next-line react/no-array-index-key
+                                        key={`${round.number}-${index}`}
```
maybe use the campaign uuid here ?

quang-le replied:
used the round id instead

beygorghor commented on plugins/polio/js/src/forms/VaccineManagementForm.tsx:
```diff
@@ -57,189 +55,119 @@ export const VaccineManagementForm: FunctionComponent<Props> = () => {
+                // className={classnames(
+                //     round.number !== currentRoundNumber &&
+                //         customClasses.hiddenOpacity,
+                // )}
```
to remove

beygorghor reviewed (changes requested):
Not an easy one this Vaccine tab ! 
See previous comment also please !

quang-le committed:
POLIO-683: review
